src/libraries/System.IO.FileSystem/tests/FileStream/ctor_str_fm_fa_fs_buffer_fo_as.cs

```diff
@@ -0,0 +1,180 @@
+            long initialAllocationSize;
+
+            using (var fs = new FileStream(filePath, mode, FileAccess.Write, FileShare.None, c_DefaultBufferSize, FileOptions.None)) // allocationSize NOT provided
+            {
+                initialAllocationSize = GetActualAllocationSize(fs); // just read it to ensure it's not being changed
+            }
+
+            using (var fs = new FileStream(filePath, mode, FileAccess.Write, FileShare.None, c_DefaultBufferSize, FileOptions.None, allocationSize))
+            {
+                Assert.Equal(initialAllocationSize, GetActualAllocationSize(fs)); // it has NOT been changed
+                Assert.Equal(initialSize, fs.Length);
+                Assert.Equal(mode == FileMode.Append ? initialSize : 0, fs.Position);
+            }
+        }
+
+        [Theory]
+        [InlineData(FileMode.Create)]
+        [InlineData(FileMode.CreateNew)]
+        [InlineData(FileMode.OpenOrCreate)]
+        public void WhenFileIsCreatedWithAllocationSizeSpecifiedTheAllocationSizeIsSet(FileMode mode)
+        {
+            const long allocationSize = 123;
+
+            using (var fs = new FileStream(GetPathToNonExistingFile(), mode, FileAccess.Write, FileShare.None, c_DefaultBufferSize, FileOptions.None, allocationSize))
+            {
+                // OS might allocate MORE than we have requested
+                Assert.True(GetActualAllocationSize(fs) >= allocationSize, $"Provided {allocationSize}, actual: {GetActualAllocationSize(fs)}");
+                Assert.Equal(GetExpectedFileLength(allocationSize), fs.Length);
+                Assert.Equal(0, fs.Position);
+            }
+        }
+
+        [OuterLoop("Might allocate 1 TB file if there is enough space on the disk")]
+        // macOS fcntl doc does not mention ENOSPC error: https://developer.apple.com/library/archive/documentation/System/Conceptual/ManPages_iPhoneOS/man2/fcntl.2.html
+        // But depending on the OS version, it might actually return it.
+        // Since we don't want to have unstable tests, it's better to not run it on macOS at all.
+        [PlatformSpecific(TestPlatforms.Windows | TestPlatforms.Linux)]
+        [Theory]
+        [InlineData(FileMode.Create)]
+        [InlineData(FileMode.CreateNew)]
+        [InlineData(FileMode.OpenOrCreate)]
+        public void WhenDiskIsFullTheErrorMessageContainsAllDetails(FileMode mode)
+        {
+            const long tooMuch = 1024L * 1024L * 1024L * 1024L; // 1 TB
+
+            string filePath = GetPathToNonExistingFile();
+
+            IOException ex = Assert.Throws<IOException>(() => new FileStream(filePath, mode, FileAccess.Write, FileShare.None, c_DefaultBufferSize, FileOptions.None, tooMuch));
+            Assert.Contains(filePath, ex.Message);
+            Assert.Contains(tooMuch.ToString(), ex.Message);
+
+            Assert.False(File.Exists(filePath)); // ensure it was NOT created (provided OOTB by Windows, emulated on Unix)
```

Comment:
Shouldn't we delete the file if it was created?

I don't know enough about `OuterLoop`, so I am assuming the environment should be reusable for future tests, in which case we should leave it as it was provided to us.

```suggestion
            // ensure it was NOT created (provided OOTB by Windows, emulated on Unix)
            bool exists = File.Exists(filePath);
            if (exists)
            {
                File.Delete(filePath);
            }
            Assert.False(exists);
```